### PR TITLE
Add draft default interface methods specification

### LINKED ISSUE #15989: Default interface implementation runtime spec
We should have it published as .md doc under https://github.com/dotnet/coreclr/tree/master/Documentation/design-docs

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Resolves #15989.

@dotnet-bot skip ci please

### CODE REVIEW

**Comment by MichalStrehovsky** [comment]
Cc @davidwrighton @AlekseyTs @marek-safar @vargaz

File: Documentation/design-docs/default-interface-methods.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    * Go over the owning types of each of the candidate methods in the list. If the owning type is less concrete than some other type in the list (there is another method in the list whose owning type requires the less concrete type), remove it from the list.
+    * If there's more than one method in the list, throw NotSupportedException
+    * If there's exactly one method in the list call that method
+    * If there's no method in the list and the interface is variant, repeat the above algorithm, looking for a variant match. Return the first variant match provided by a most specific interface.
+
+**Section** "III.2.1 constrained. prefix" the paragraph starting with "This last case can only occur when method was defined on `System.Object`, `System.ValueType`, or `System.Enum`" is extended to also cover default interface method implementation. In the case the interface method implementation is provided by an interface, the implicit boxing becomes _observable_ to the program.
+
+**Section** "III.4.2 callvirt" is extended to allow throwing `NotSupportedException` if the implementation of the interface method resolves at runtime to more than one default interface method.
```
[the remark this answers is not in the document]
Comment (MichalStrehovsky) [reply Documentation/design-docs/default-interface-methods.md]: Pending dotnet/corefx#34124

[the remark this answers is not in the document]
Comment (davidwrighton) [reply Documentation/design-docs/default-interface-methods.md]: This looks good to me, but I'd wait on the outcome of corefx 34124 before I sign off.

[the remark this answers is not in the document]
Comment (davidwrighton) [reply Documentation/design-docs/default-interface-methods.md]: This looks good to me, but I'd wait on the outcome of corefx 34124 before I sign off.

**Pull request merged**